src/helpers/performance.js

```diff
@@ -0,0 +1,44 @@
+const performance = typeof window !== "undefined" ? window.performance : null;
+
+export const hasHighPrecision = Boolean(performance);
+
+// Safari doesn't expose the performance API, so make a wrapper API and expose a
+// lower precision form of it if not available.
+let clear = (perfMark) => {
+  performance.clearMarks(perfMark);
+  performance.clearMeasures(perfMark);
+};
+let internalMark = performance?.mark.bind(performance);
+let internalMeasure = (perfMark) => {
+  const { duration } = performance.measure(null, perfMark);
+  clear(perfMark);
+  return { time: duration, hasHighPrecision };
+};
+if (!hasHighPrecision) {
+  const marks = {};
+  internalMark = (perfMark) => {
+    marks[perfMark] = Date.now();
+  };
+  clear = (perfMark) => {
+    marks[perfMark] = undefined;
```

Comment:
this may not be a possible use case, but should this be set to `0` on `clear` since it might possibly be subtracted from `Date.now()` in `internalMeasure`

Reply:
Hm fair point, but that would only break if `mark` isn't called first, which would break before `clear` is called as well. I'm okay leaving this as is for now but if we end up stealing this in future projects it's something to consider.